### PR TITLE
Add support for Responsible AI

### PR DESCRIPTION
**Pull request by jdubois**
This adds support for Responsible AI (content filtered by Azure OpenAI for violence, self harm, or hate).

This is a fix for #688 

- Support for filtered content in the models
- Test cases to validate this support

Warning: the test cases do give examples of violence and self harm, this is why they are in a separate class, with a warning on top of it.

## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling across various Azure OpenAI models to improve user feedback on failed operations.
- **Refactor**
	- Improved logging for error handling in Azure OpenAI image model generation.
- **Tests**
	- Added tests for responsible AI behavior, including content filtering for violence, self-harm, and sexual content across chat messages, images, and language models.

### CODE REVIEW

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiImageModel.java [hunk 1]
```diff
@@ -150,9 +159,34 @@ public Response<Image> generate(String prompt) {
...
 
-        ImageGenerations imageGenerations = client.getImageGenerations(deploymentName, options);
-        Image image = imageFrom(imageGenerations.getData().get(0));
-        return Response.from(image);
+        try {
+            ImageGenerations imageGenerations = client.getImageGenerations(deploymentName, options);
+            Image image = imageFrom(imageGenerations.getData().get(0));
+            return Response.from(image);
+        } catch (HttpResponseException httpResponseException) {
+            logger.info("Error generating image, {}", httpResponseException.getValue());
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiImageModel.java:166]: Please extract this block into util method, to reduce code duplication here and in all other classes

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModel.java [hunk 2]
```diff
@@ -244,13 +252,41 @@ private Response<AiMessage> generate(List<ChatMessage> messages,
...
+                            }
+                        }
+                    }
+                } catch (ClassCastException classCastException) {
+                    logger.error("Error parsing error response from Azure OpenAI", classCastException);
+                }
+            }
+            return Response.from(
+                    aiMessage(exceptionMessage),
+                    new TokenUsage(),
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModel.java]:
```suggestion
                    null,
```

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiImageModel.java [hunk 3]
```diff
@@ -150,9 +159,34 @@ public Response<Image> generate(String prompt) {
...
+                                // The content was filtered by Azure OpenAI's content filter (for violence, self harm, or hate).
+                                exceptionFinishReason = FinishReason.CONTENT_FILTER;
+                            }
+                        }
+                    }
+                } catch (ClassCastException classCastException) {
+                    logger.error("Error parsing error response from Azure OpenAI", classCastException);
+                }
+            }
+            return Response.from(Image.builder().build(), new TokenUsage(), exceptionFinishReason);
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiImageModel.java]:
```suggestion
            return Response.from(Image.builder().build(), null, exceptionFinishReason);
```

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiLanguageModel.java [hunk 4]
```diff
@@ -213,13 +221,41 @@ public Response<String> generate(String prompt) {
...
+                            }
+                        }
+                    }
+                } catch (ClassCastException classCastException) {
+                    logger.error("Error parsing error response from Azure OpenAI", classCastException);
+                }
+            }
+            return Response.from(
+                    exceptionMessage,
+                    new TokenUsage(),
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiLanguageModel.java]:
```suggestion
                    null,
```

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingChatModel.java [hunk 5]
```diff
@@ -269,6 +277,34 @@ private void generate(List<ChatMessage> messages,
...
+                            }
+                        }
+                    }
+                } catch (ClassCastException classCastException) {
+                    logger.error("Error parsing error response from Azure OpenAI", classCastException);
+                }
+            }
+            Response<AiMessage> response = Response.from(
+                    aiMessage(exceptionMessage),
+                    new TokenUsage(),
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingChatModel.java]:
```suggestion
                    null,
```

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingLanguageModel.java [hunk 6]
```diff
@@ -224,6 +232,33 @@ public void generate(String prompt, StreamingResponseHandler<String> handler) {
...
+                            }
+                        }
+                    }
+                } catch (ClassCastException classCastException) {
+                    logger.error("Error parsing error response from Azure OpenAI", classCastException);
+                }
+            }
+            handler.onComplete(Response.from(
+                    exceptionMessage,
+                    new TokenUsage(),
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingLanguageModel.java]:
```suggestion
                    null,
```

**Review by ghost — COMMENTED** [review]
@jdubois looks good, thank you! Just a few minor comments

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiImageModel.java (hunk 1 above)
Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiImageModel.java:166]: Indeed! Done.

**Comment by jdubois** [comment]
Thanks @langchain4j ! This should be OK now.

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModel.java [hunk 7]
```diff
@@ -244,13 +252,22 @@ private Response<AiMessage> generate(List<ChatMessage> messages,
...
-        );
+        try {
+            ChatCompletions chatCompletions = client.getChatCompletions(deploymentName, options);
+            return Response.from(
+                    aiMessageFrom(chatCompletions.getChoices().get(0).getMessage()),
+                    tokenUsageFrom(chatCompletions.getUsage()),
+                    finishReasonFrom(chatCompletions.getChoices().get(0).getFinishReason())
+            );
+        } catch (HttpResponseException httpResponseException) {
+            logger.info("Error generating response, {}", httpResponseException.getValue());
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModel.java:263]: minor: is logging required in this case? User will know about content filter from the returned `Response`
(this applies to all classes)

**Review by ghost — APPROVED** [review]
@jdubois thanks a lot!

**Comment by langchain4j** [comment]
Hi @jdubois, some tests in `AzureOpenAIResponsibleAIIT` are failing on CI very often:

```
2024-10-08T00:15:51.9341760Z [ERROR]   AzureOpenAIResponsibleAIIT.chat_message_should_trigger_content_filter_for_self_harm:71 
2024-10-08T00:15:51.9343038Z Expecting actual:
2024-10-08T00:15:51.9345131Z   "I'm really sorry you're feeling this way, but I'm unable to help. It's important to talk to a mental health professional or someone you trust about what you're going through."
2024-10-08T00:15:51.9346611Z to contain:
2024-10-08T00:15:51.9347252Z   ["ResponsibleAIPolicyViolation",
2024-10-08T00:15:51.9348167Z     ""self_harm":{"filtered":true,"severity":"medium"}"]
2024-10-08T00:15:51.9349041Z but could not find:
2024-10-08T00:15:51.9349856Z   ["ResponsibleAIPolicyViolation",
2024-10-08T00:15:51.9350491Z     ""self_harm":{"filtered":true,"severity":"medium"}"]
2024-10-08T00:15:51.9351015Z  
2024-10-08T00:15:51.9351645Z [ERROR]   AzureOpenAIResponsibleAIIT.chat_message_should_trigger_content_filter_for_violence:51 
2024-10-08T00:15:51.9352375Z Expecting actual:
2024-10-08T00:15:51.9356841Z   "I'm really sorry that you're feeling this way, but I'm unable to help. It's important to treat all living beings with kindness and respect. If you're struggling with unwanted kittens or any other issues, please consider reaching out to a local animal shelter or rescue organization for assistance. They can provide support and resources to help you find a humane solution. If you're feeling overwhelmed or distressed, talking to a mental health professional can also be very beneficial."
2024-10-08T00:15:51.9361486Z to contain:
2024-10-08T00:15:51.9362002Z   ["ResponsibleAIPolicyViolation",
2024-10-08T00:15:51.9362751Z     ""violence":{"filtered":true,"severity":"medium"}"]
2024-10-08T00:15:51.9363400Z but could not find:
2024-10-08T00:15:51.9363977Z   ["ResponsibleAIPolicyViolation",
2024-10-08T00:15:51.9364633Z     ""violence":{"filtered":true,"severity":"medium"}"]
2024-10-08T00:15:51.9365140Z  
2024-10-08T00:15:51.9366382Z [ERROR]   AzureOpenAIResponsibleAIIT.streaming_chat_message_should_trigger_content_filter_for_violence:152 
2024-10-08T00:15:51.9367626Z Expecting actual:
2024-10-08T00:15:51.9372355Z   "I'm really sorry that you're feeling this way, but I can't assist with that. It's important to treat all living beings with kindness and respect. If you're struggling with unwanted animals or feeling overwhelmed, please reach out to a local animal shelter or a professional for guidance. If you need someone to talk to, consider speaking with a mental health professional."
2024-10-08T00:15:51.9374877Z to contain:
2024-10-08T00:15:51.9375280Z   ["ResponsibleAIPolicyViolation",
2024-10-08T00:15:51.9376159Z     ""violence":{"filtered":true,"severity":"medium"}"]
2024-10-08T00:15:51.9376834Z but could not find:
2024-10-08T00:15:51.9377382Z   ["ResponsibleAIPolicyViolation",
2024-10-08T00:15:51.9377997Z     ""violence":{"filtered":true,"severity":"medium"}"]
```

Should I remove/disable the failing ones (the rest works fine and still tests this feature) or do you want to tinker with prompts to make it work?

**Comment by jdubois** [comment]
Oh that's probably because the filters were refined.... I tried to find some that fail but which aren't too bad.
Let me have a look at having "better" prompts.

**Comment by jdubois** [comment]
So it's not the filter, that's the GPT-4o model which is more "responsible". That's good to have better models, but now it's inconsistent with the filters as they don't trigger those policy violations anymore!!
I'm guessing the easiest solution is to use a model which is bad at this -> I might migrate this test to GitHub Models, which should use the same filtering system, and which has access to more models, and non-OpenAI or non-Microsoft models, so probably they will be less "responsible".

**Comment by langchain4j** [comment]
@jdubois I see. Ideally we should have this test for both Azure OpenAI and Github Models. As far as I understand, the rest of the tests still work and test this feature, so we could probably keep the smaller test suite (only passing tests) for Azure OpenAI and complete test suite in Github Models?

**Comment by jdubois** [comment]
That's the same system for both, but as the SDKs are different I'd rather have some good tests for both.

This is really annoying on our side:
- As those APIs are changing often (they just added some new types of policies), I'd rather do integration tests and not mocks
- The new OpenAI models are so good that I can't reliably trigger the filters anymore

First thing would be to migrate to GitHub Models and use some non-OpenAI models, like the Mistral ones, as last time I checked it was super easy to make them do the wrong thing. Then I'm not sure how we can have something for Azure OpenAI... I guess it's better than what we actually have, so that would still be an improvement.

**Comment by langchain4j** [comment]
> That's the same system for both, but as the SDKs are different I'd rather have some good tests for both.
> This is really annoying on our side:
> As those APIs are changing often (they just added some new types of policies), I'd rather do integration tests and not mocks

👍 


> Then I'm not sure how we can have something for Azure OpenAI... I guess it's better than what we actually have, so that would still be an improvement.

What do you mean?

**Comment by jdubois** [comment]
What I mean is that it's currently not working with Azure OpenAI as the models have been reinforced to prevent harm, so it's really hard to hit the "responsible AI" filters.
So those tests are now quite useless for Azure OpenAI.
As it's the same system for GitHub Models, but that they have models without this reinforcement (like the Mistral ones), testing the filters there will be much more meaningful than what we currently have.

**Comment by langchain4j** [comment]
@jdubois got it, thanks! So what are the action points? I would really love to get rid of everyday failures of Azure OpenAI module

**Comment by glaforge** [comment]
Do they ever succeed?
If yes, maybe adding a retry able test annotation would get a good chance to have them pass?

**Comment by jdubois** [comment]
@glaforge they used to!! This is surprising for a lot of people, as the new models are a lot better at managing this than before. Yes we should have a retry, but we need to migrate to GitHub Models first, are those OpenAI models are so good that even with the retry it won't be enough... I managed to have 1 test succeeding in about 10 tries, with a prompt like "write a short horror story" (the trick is not to ask directly the model for something bad, but push it to write it, so that the filter will get it).

**Comment by jdubois** [comment]
I just did https://github.com/langchain4j/langchain4j/pull/1945 to try to fix this, please see my comments there to have the details

**Pull request merged**